### ISSUE TITLE
Arguments of parametrized types are not properly checked

### ISSUE THREAD

**Issue by mikir** [ndsev/zserio#157]
Argument types of parametrized types are not checked properly for compound and enumeration types. There is only check if the argument type is compound or enumeration but no further check for concrete instance of compound or enumeration type is performed.

Example which compiles without any problem:

```
struct Param1
{
    int32   field;
};

struct Param2
{
    bool    something;
};

struct ParametrizedType(Param1 param1)
{
    int32   field;
    int32   hasAnother if param1.field != 0;
};

struct Tester
{
    Param1                   param1;
    ParametrizedType(param1) parametrizedType1;
    Param2                   param2;
    ParametrizedType(param2) parametrizedType2;
};
```

**Comment by Mi-La** [comment]
```
struct ArrayHolder(uint32 size)
{
    uint32 array[size];
};
struct Test
{
    ArrayHolder(10) arrayHolder;
};
```
Note that parameter `10` forced by `Test` structure is never check during write and is only used during read.

**Comment by mikir** [comment]
> ```
> struct ArrayHolder(uint32 size)
> {
>     uint32 array[size];
> };
> struct Test
> {
>     ArrayHolder(10) arrayHolder;
> };
> ```
> 
> Note that parameter `10` forced by `Test` structure is never check during write and is only used during read.

This will be solved within new issue #378.

**Comment by mikir** [comment]
> A more complete example which demonstrates one of the problems. This behavior can lead to errors which are detected quite late if you don't do a round trip check:
> 
> ```c++
> package myschema;
> 
> struct Outer             
> {                        
>     Inner(5) value;      
> };                       
>                          
> struct Inner(int numBits)
> {                        
>   int<numBits> value;    
> };                       
> ```
> 
> ```python
> # Test the generated API here:
> from myschema.api import *    
> import zserio                 
> 
> test = Outer(value_=Inner(num_bits_=10, value_=10))                                       
> print(f"ORIGINAL OBJECT:    num_bits: {test.value.num_bits} value: {test.value.value}");  
>                                                                                           
> bytes = zserio.serialize_to_bytes(test)                                                   
> test2 = zserio.deserialize_bytes(Outer, bytes)                                            
> print(f"DESERIALIZE RESULT: num_bits: {test2.value.num_bits} value: {test2.value.value}");
> ```
> 
> ```
> ORIGINAL OBJECT:    num_bits: 10 value: 10
> DESERIALIZE RESULT: num_bits: 5 value: 0
> ```

This will be solved within new issue #378.
